keyboards/matrix/falcon_magnet/apc.c

```diff
@@ -0,0 +1,222 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "apc.h"
+
+#define APC_KEY_STATE_DEFAULT   APC_KEY_OFF
+#define APC_KEY_MIN_DEFAULT     APC_KEY_MAX
+#define APC_KEY_MAX_DEFAULT     APC_KEY_MIN
+#define APC_KEY_LAST_DEFAULT    0
+#define APC_KEY_TRIGGER_DEFAULT 0
+#define APC_KEY_DOWN_DEFAULT    1024
+#define APC_KEY_UP_DEFAULT      0
+
+enum apc_key_state
+{
+    APC_KEY_OFF,
+    APC_KEY_PRESSING,
+    APC_KEY_ON,
+    APC_KEY_RELEASING,
+};
+
+struct apc_key
+{
+    uint32_t state;
+    uint32_t min;
+    uint32_t max;
+    uint32_t last;
+    uint32_t trigger;
+    uint32_t down;
+    uint32_t up;
+    bool on;
+};
+
+static struct apc_key apc_matrix[MATRIX_ROWS][MATRIX_COLS];
+
+#ifdef DYNAMIC_KEYMAP_ENABLE
+#include "dynamic_keymap.h"
+#else
+extern const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS];
+#endif
+
+//
+// TODO:
+// actually the relationship between the interval and the magnetic value
+// was non-linear. Currently implementation was not accurate but can work.
+//
+static uint32_t apc_compute_interval(uint32_t row, uint32_t col, uint32_t index)
+{
+    struct apc_key *key = &apc_matrix[row][col];
+    uint32_t interval = APC_INTERVAL_INVALID;
+
+    if ((key->min == APC_KEY_MIN_DEFAULT) || (key->max < APC_KEY_MAX)) {
+        interval = APC_INTERVAL_MIN + ((APC_INTERVAL_MAX-APC_INTERVAL_MIN)/APC_INTERVAL_COUNT)*(index);
+    } else {
+        interval = APC_INTERVAL_MIN + ((key->max-key->min)/APC_INTERVAL_COUNT)*(index);
+    }
+
+    return interval;
+}
+
+// currently two keymap was use for set the down&up actuation point. the up actuation also
+// enabled the rapid trigger feature
+static uint32_t apc_get_key_interval(uint32_t row, uint32_t col, uint32_t layer)
+{
+#ifdef DYNAMIC_KEYMAP_ENABLE
+    uint16_t keycode = dynamic_keymap_get_keycode(layer, row, col);
+#else
+    uint16_t keycode = keymaps[layer][row][col];
+#endif
+    uint16_t index = 0;
+    if (keycode>=KC_F1 && keycode <= KC_F12) {
+        index = keycode - KC_F1 + 1;
+    } else if (keycode>=KC_F13 && keycode <= KC_F24) {
+        index = keycode - KC_F13 + (KC_F12 - KC_F1) + 1;
+    } else {
+        if (layer == APC_KEYMAP_UP_LAYER) {
+            return APC_INTERVAL_INVALID;
+        }
+        index = APC_INTERVAL_INDEX;
+    }
+
+    return apc_compute_interval(row, col, index);
+}
+
+static void apc_update_key_interval(uint32_t row, uint32_t col)
+{
+    apc_matrix[row][col].down = apc_get_key_interval(row, col, APC_KEYMAP_DOWN_LAYER);
+    apc_matrix[row][col].up = apc_get_key_interval(row, col, APC_KEYMAP_UP_LAYER);
+}
+
+void apc_matrix_init(void)
+{
+    for (int i = 0; i < MATRIX_ROWS; i++) {
+        for (int j = 0; j < MATRIX_COLS; j++) {
```

Comment:
is a nitpick, but you could use uint8_t instead of int, for both of this.

Reply:
done